Define a shell script for building our Dockerfile

krzysz00 opened the pull request:
Currently, the Dockerfile build is inline in a Jenkins job config,
which makes it impossible to track changes and. Move the logic from
that build pipeline (with minor refactoring) to a script that the
Jenkins job can call.

This PR will have no effect until the rocm-docker job on ml-ci.amd.com
is changed to use the new script.

krzysz00 committed:
Define a shell script for building our Dockerfile

Currently, the Dockerfile build is inline in a Jenkins job config,
which makes it impossible to track changes and. Move the logic from
that build pipeline (with minor refactoring) to a script that the
Jenkins job can call.

This PR will have no effect until the rocm-docker job on ml-ci.amd.com
is changed to use the new script.

jerryyin committed:
Amend `rocm_full_version` to parse Dockerfile.rocm

jerryyin commented on mlir/utils/jenkins/build-dockerfile.sh:
```diff
@@ -0,0 +1,68 @@
+}
+
+function main() {
+  if git diff --name-only HEAD^ HEAD | grep -q "Dockerfile.rocm"; then
```
[Nit] I'd prefer to have a jenkins end controllable option to skip the dependency check. This means: When a build failed (and it happened from case to case), we have a way to force another build without having to change the source.

jerryyin reviewed (approved):
Looks good. Before we transition to this method, I'd prefer to have the [Nit] addressed so there's more control over the jenkins end.

krzysz00 committed:
Add a "force" option

Since the Jenkins job isn't set up as a pipeline, we can't add
parameters, but we can make the script take an argument to make all
this easier.

manupak commented on mlir/utils/jenkins/build-dockerfile.sh:
```diff
@@ -0,0 +1,71 @@
+# Note: includes a workaround for the occasional hung docker daemon
+start_clean_docker() {
+  systemctl status docker | grep 'Active:'
+  sudo /usr/bin/pkill -f docker
```
@krzysz00 what was the intention behind this line ? was it to stop and remove the running containers ?
